File: packages/tale_knowledge/src/tale_knowledge/embedding/service.py
"""OpenAI-compatible embedding generation service.

Constructor-injected configuration — no global state or settings imports.
Each service creates its own EmbeddingService instance with its own config.
"""

import asyncio
import random
from dataclasses import dataclass, field

from loguru import logger
from openai import (
    APIConnectionError,
    APITimeoutError,
    AsyncOpenAI,
    InternalServerError,
    RateLimitError,
)

MAX_BATCH_SIZE = 256
MAX_CONCURRENT_REQUESTS = 3
MAX_RETRIES = 3
RETRY_BASE_DELAY = 1.0


@dataclass
class EmbeddingUsage:
    """Token usage from embedding API calls."""

    prompt_tokens: int = 0
    total_tokens: int = 0
    model: str = ""

    def add(self, prompt: int, total: int) -> None:
        self.prompt_tokens += prompt
        self.total_tokens += total

    def to_dict(self) -> dict:
        return {
            "prompt_tokens": self.prompt_tokens,
            "total_tokens": self.total_tokens,
            "model": self.model,
        }


@dataclass
class EmbeddingResult:
    """Embeddings with token usage metadata."""

    embeddings: list[list[float]] = field(default_factory=list)
    usage: EmbeddingUsage = field(default_factory=EmbeddingUsage)


@dataclass
class EmbeddingQueryResult:
    """Single embedding with token usage metadata."""

    embedding: list[float] = field(default_factory=list)
    usage: EmbeddingUsage = field(default_factory=EmbeddingUsage)


class EmbeddingService:
    def __init__(self, api_key: str, base_url: str | None, model: str, dimensions: int):
        self._client = AsyncOpenAI(api_key=api_key, base_url=base_url)
        self._model = model
        self._dimensions = dimensions
        self._semaphore = asyncio.Semaphore(MAX_CONCURRENT_REQUESTS)

    @property
    def dimensions(self) -> int:
        return self._dimensions

    def _zero_vector(self) -> list[float]:
        return [0.0] * self._dimensions

    async def _embed_batch_with_usage(self, batch: list[str], usage: EmbeddingUsage) -> list[list[float]]:
        valid = [(i, text) for i, text in enumerate(batch) if text.strip()]
        if not valid:
            return [self._zero_vector() for _ in batch]

        valid_indices, valid_texts = zip(*valid, strict=True)

        async with self._semaphore:
            for attempt in range(MAX_RETRIES):
                try:
                    response = await self._client.embeddings.create(
                        model=self._model,
                        input=list(valid_texts),
                        dimensions=self._dimensions,
                    )
                    if not response.data:
                        raise ValueError("No embedding data received")
                    embeddings = [item.embedding for item in response.data]
                    if response.usage:
                        usage.add(
                            response.usage.prompt_tokens,
                            response.usage.total_tokens,
                        )
                    break
                except (
                    RateLimitError,
                    APITimeoutError,
                    APIConnectionError,
                    InternalServerError,
                ):
                    if attempt == MAX_RETRIES - 1:
                        raise
                    delay = RETRY_BASE_DELAY * (2**attempt) + random.uniform(0, 0.5)
                    logger.warning(
                        "Embedding request failed (attempt {}/{}), retrying in {:.2f}s",
                        attempt + 1,
                        MAX_RETRIES,
                        delay,
                    )
                    await asyncio.sleep(delay)
                except ValueError:
                    logger.warning(
                        "Embedding returned empty data for batch of {} texts, filling with zero vectors",
                        len(valid_texts),
                    )
                    return [self._zero_vector() for _ in batch]

        results: list[list[float]] = [self._zero_vector() for _ in batch]
        for idx, emb in zip(valid_indices, embeddings, strict=True):
            results[idx] = emb
        return results

    async def _embed_batch(self, batch: list[str]) -> list[list[float]]:
        return await self._embed_batch_with_usage(batch, EmbeddingUsage())

    async def embed_texts(self, texts: list[str]) -> list[list[float]]:
        result = await self.embed_texts_with_usage(texts)
        return result.embeddings

    async def embed_texts_with_usage(self, texts: list[str]) -> EmbeddingResult:
        if not texts:
            return EmbeddingResult()
        usage = EmbeddingUsage(model=self._model)
        batches = [texts[i : i + MAX_BATCH_SIZE] for i in range(0, len(texts), MAX_BATCH_SIZE)]
        results = await asyncio.gather(*[self._embed_batch_with_usage(batch, usage) for batch in batches])
        embeddings = [emb for batch_result in results for emb in batch_result]
        return EmbeddingResult(embeddings=embeddings, usage=usage)

    async def embed_query(self, query: str) -> list[float]:
        result = await self.embed_query_with_usage(query)
        return result.embedding

    async def embed_query_with_usage(self, query: str) -> EmbeddingQueryResult:
        result = await self.embed_texts_with_usage([query])
        return EmbeddingQueryResult(
            embedding=result.embeddings[0] if result.embeddings else self._zero_vector(),
            usage=result.usage,
        )

    async def close(self):
        await self._client.close()

File: services/platform/convex/lib/knowledge/embedding/service.ts
'use node';

/**
 * OpenAI-compatible embedding generation service.
 *
 * Constructor-injected configuration — no global state or settings imports.
 * Each service creates its own EmbeddingService instance with its own config.
 */

import OpenAI, {
  APIConnectionError,
  APIConnectionTimeoutError,
  InternalServerError,
  RateLimitError,
} from 'openai';

export const MAX_BATCH_SIZE = 256;
export const MAX_CONCURRENT_REQUESTS = 3;
export const MAX_RETRIES = 3;
export const RETRY_BASE_DELAY_MS = 1000;

export interface EmbeddingUsage {
  promptTokens: number;
  totalTokens: number;
  model: string;
}

export interface EmbeddingResult {
  embeddings: number[][];
  usage: EmbeddingUsage;
}

export interface EmbeddingQueryResult {
  embedding: number[];
  usage: EmbeddingUsage;
}

function emptyUsage(model = ''): EmbeddingUsage {
  return { promptTokens: 0, totalTokens: 0, model };
}

function isRetryable(err: unknown): boolean {
  return (
    err instanceof RateLimitError ||
    err instanceof APIConnectionTimeoutError ||
    err instanceof APIConnectionError ||
    err instanceof InternalServerError
  );
}

const sleep = (ms: number): Promise<void> =>
  new Promise((r) => setTimeout(r, ms));

export class EmbeddingService {
  protected client: OpenAI;
  private readonly model: string;
  private readonly dims: number;
  private active = 0;
  private readonly waiters: (() => void)[] = [];

  constructor(
    apiKey: string,
    baseUrl: string | null,
    model: string,
    dimensions: number,
  ) {
    this.client = new OpenAI({ apiKey, baseURL: baseUrl ?? undefined });
    this.model = model;
    this.dims = dimensions;
  }

  /**
   * Replace the underlying OpenAI client. Intended for tests to inject a stub
   * with a mocked `embeddings.create`; production code never calls this.
   */
  setClient(client: OpenAI): void {
    this.client = client;
  }

  get dimensions(): number {
    return this.dims;
  }

  private zeroVector(): number[] {
    return new Array<number>(this.dims).fill(0.0);
  }

  /** A small concurrency gate mirroring the Python asyncio.Semaphore(3). */
  private async acquire(): Promise<void> {
    if (this.active < MAX_CONCURRENT_REQUESTS) {
      this.active += 1;
      return;
    }
    await new Promise<void>((resolve) => this.waiters.push(resolve));
    this.active += 1;
  }

  private release(): void {
    this.active -= 1;
    const next = this.waiters.shift();
    if (next) {
      next();
    }
  }

  private async embedBatchWithUsage(
    batch: string[],
    usage: EmbeddingUsage,
  ): Promise<number[][]> {
    const valid: [number, string][] = [];
    batch.forEach((text, i) => {
      if (text.trim()) {
        valid.push([i, text]);
      }
    });
    if (valid.length === 0) {
      return batch.map(() => this.zeroVector());
    }

    const validIndices = valid.map(([i]) => i);
    const validTexts = valid.map(([, text]) => text);

    await this.acquire();
    try {
      let embeddings: number[][] = [];
      for (let attempt = 0; attempt < MAX_RETRIES; attempt += 1) {
        try {
          const response = await this.client.embeddings.create({
            model: this.model,
            input: validTexts,
            dimensions: this.dims,
          });
          if (!response.data || response.data.length === 0) {
            console.warn(
              `Embedding returned empty data for batch of ${validTexts.length} texts, ` +
                `filling with zero vectors`,
            );
            return batch.map(() => this.zeroVector());
          }
          embeddings = response.data.map((item) => item.embedding);
          if (response.usage) {
            usage.promptTokens += response.usage.prompt_tokens;
            usage.totalTokens += response.usage.total_tokens;
          }
          break;
        } catch (err) {
          if (!isRetryable(err) || attempt === MAX_RETRIES - 1) {
            throw err;
          }
          const delay =
            RETRY_BASE_DELAY_MS * 2 ** attempt + Math.random() * 500;
          console.warn(
            `Embedding request failed (attempt ${attempt + 1}/${MAX_RETRIES}), ` +
              `retrying in ${(delay / 1000).toFixed(2)}s`,
          );
          await sleep(delay);
        }
      }

      const results: number[][] = batch.map(() => this.zeroVector());
      validIndices.forEach((idx, j) => {
        results[idx] = embeddings[j];
      });
      return results;
    } finally {
      this.release();
    }
  }

  async embedTexts(texts: string[]): Promise<number[][]> {
    return (await this.embedTextsWithUsage(texts)).embeddings;
  }

  async embedTextsWithUsage(texts: string[]): Promise<EmbeddingResult> {
    if (texts.length === 0) {
      return { embeddings: [], usage: emptyUsage() };
    }
    const usage = emptyUsage(this.model);
    const batches: string[][] = [];
    for (let i = 0; i < texts.length; i += MAX_BATCH_SIZE) {
      batches.push(texts.slice(i, i + MAX_BATCH_SIZE));
    }
    const results = await Promise.all(
      batches.map((batch) => this.embedBatchWithUsage(batch, usage)),
    );
    return { embeddings: results.flat(), usage };
  }

  async embedQuery(query: string): Promise<number[]> {
    return (await this.embedQueryWithUsage(query)).embedding;
  }

  async embedQueryWithUsage(query: string): Promise<EmbeddingQueryResult> {
    const result = await this.embedTextsWithUsage([query]);
    return {
      embedding:
        result.embeddings.length > 0 ? result.embeddings[0] : this.zeroVector(),
      usage: result.usage,
    };
  }
}
